File: transit-play/app/controllers/Application.java
package controllers;

import com.weizilla.transit.bus.BusController;
import com.weizilla.transit.bus.data.Direction;
import com.weizilla.transit.bus.data.Prediction;
import com.weizilla.transit.bus.data.Route;
import com.weizilla.transit.bus.data.Stop;
import com.weizilla.transit.bus.source.stream.StreamingBusDataSource;
import com.weizilla.transit.bus.source.stream.http.HttpInputStreamProvider;
import com.weizilla.transit.bus.source.stream.http.HttpReader;
import com.weizilla.transit.favorites.sqlite.SqliteFavoritesStore;
import play.mvc.Controller;
import play.mvc.Result;
import views.html.*;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collection;

public class Application extends Controller
{
    private static final Path favDbPath = createPath();

    private static Path createPath()
    {
        try
        {
            Path favDbPath = Files.createTempFile("transit-play-", ".db");
            favDbPath.toFile().deleteOnExit();
            return favDbPath;
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }
        return null;
    }

    private static BusController createController()
    {
        String apiKey = "vqfwTE6FQG778wgBkcy5vfFgh";
        HttpReader reader = new HttpReader(new HttpReader.HttpURLConnectionFactory());
        HttpInputStreamProvider provider = new HttpInputStreamProvider(reader, apiKey);
        StreamingBusDataSource source = new StreamingBusDataSource(provider);

        SqliteFavoritesStore store = null;
        try
        {
            store = SqliteFavoritesStore.createStore(favDbPath);
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }

        return new BusController(source, store);
    }

    public static Result index()
    {
        BusController controller = createController();
        Collection<Route> routes = controller.getRoutes();
        Collection<String> favRoutes = controller.getFavoriteRoutes();
        return ok(index.render("Bus - routes",  routes, favRoutes));
    }

    public static Result directions(String routeId)
    {
        //TODO generate route here using just id?
        Route route = new Route(routeId);
        Collection<Direction> dirs = createController().getDirections(route);
        return ok(directions.render("Bus - directions", route, dirs));
    }

    public static Result stops(String routeId, String direction)
    {
        BusController controller = createController();
        //TODO generate route here using just id?
        Route route = new Route(routeId);
        Direction dir = Direction.valueOf(direction);
        Collection<Stop> allStops = controller.getStops(route, dir);

        Collection<Integer> favStopIds = controller.getFavoriteStops(routeId, dir);
        Collection<Stop> favStops = new ArrayList<>(favStopIds.size());
        for (Stop stop : allStops)
        {
            if (favStopIds.contains(stop.getId()))
            {
                favStops.add(stop);
            }
        }
        return ok(stops.render("Bus - stops", route, allStops, favStops));
    }

    public static Result predictions(String routeId, int stopId)
    {
        //TODO generate route here using just id?
        Route route = new Route(routeId);
        Stop stop = new Stop(stopId);
        Collection<Prediction> prds = createController().getPredictions(route, stop);
        if (prds == null)
        {
            return ok(message.render("Bus - predictions", "No predictions"));
        }
        else {
            return ok(predictions.render("Bus - predictions", prds));
        }
    }

    public static Result saveFavoriteRoute(String routeId)
    {
        BusController controller = createController();
        controller.saveFavorite(new Route(routeId));
        return ok(message.render("Bus - Favorite", "Route " + routeId + " saved as favorite"));
    }

    public static Result saveFavoriteStop(String routeId, String direction, int stopId)
    {
        BusController controller = createController();
        //TODO create stop here?
        Stop stop = new Stop(stopId);
        stop.setDirection(Direction.valueOf(direction));
        stop.setRouteId(routeId);
        controller.saveFavorite(stop);
        return ok(message.render("Bus - Favorite", "Stop " + stopId + " saved as favorite"));
    }
}

File: transit-play/app/controllers/Application.scala
package controllers

import java.nio.file.Files

import com.typesafe.config.ConfigFactory
import com.weizilla.transit.BusController
import com.weizilla.transit.data.{Direction, Prediction}
import com.weizilla.transit.favorites.sqlite.JdbcSqliteFavoritesStore
import com.weizilla.transit.groups.sqlite.JdbcSqliteGroupsStore
import com.weizilla.transit.cache.sqlite.JdbcSqliteCacheStore
import com.weizilla.transit.source.stream.StreamingDataSource
import com.weizilla.transit.source.stream.http.{HttpInputStreamProvider, HttpReader}
import play.api.data.Form
import play.api.data.Forms._
import play.api.mvc.{Action, Controller}

import scala.collection.JavaConversions._

object Application extends Controller {
  val apiKey = ConfigFactory.load("api-key.conf").getString("apiKey")

  val controller = {
    val reader: HttpReader = new HttpReader(new HttpReader.HttpURLConnectionFactory)
    val provider: HttpInputStreamProvider = new HttpInputStreamProvider(reader, apiKey)
    val source: StreamingDataSource = new StreamingDataSource(provider)

    try {
      val dbPath = Files.createTempFile("transit-play-", ".db")
      dbPath.toFile.deleteOnExit()
      val favStore: JdbcSqliteFavoritesStore = JdbcSqliteFavoritesStore.createStore(dbPath)
      val groupsStore: JdbcSqliteGroupsStore = JdbcSqliteGroupsStore.createStore(dbPath)
      val cacheStore: JdbcSqliteCacheStore = JdbcSqliteCacheStore.createStore(dbPath)
      new BusController(source, favStore, groupsStore, cacheStore)
    } catch {
      case e: Exception =>
        throw new RuntimeException("Unable to create bus controller", e)
    }
  }

  val createGroupForm = Form(single("name" -> nonEmptyText))

  val addStopToGroupForm = Form(tuple(
    "stopId" -> number,
    "groupId" -> number,
    "routeId" -> nonEmptyText,
    "direction" -> nonEmptyText
  ))

  val saveFavoriteRouteForm = Form(single("routeId" -> nonEmptyText))
  val saveFavoriteStopForm = Form(tuple(
    "stopId" -> number,
    "routeId" -> nonEmptyText,
    "direction" -> nonEmptyText
  ))

  def index = Action {
    val groups = controller.getAllGroups
    val routes = controller.getRoutes
    val routeMap = routes.map(r => r.getId -> r).toMap
    val favRoutes = controller.getFavoriteRouteIds.map(routeMap(_))
    Ok(views.html.index(groups, favRoutes, routes))
  }

  def directions(routeId: String) = Action {
    val directions = controller.getDirections(routeId)
    Ok(views.html.directions(routeId, directions))
  }

  def stops(routeId: String, direction: String) = Action {
    val dir = Direction.valueOf(direction)
    val stops = controller.getStops(routeId, dir)
    val stopsMap = stops.map(s => s.getId -> s).toMap
    val favStops = controller.getFavoriteStopIds.map(stopsMap(_))
    val groups = controller.getAllGroups
    Ok(views.html.stops(routeId, direction, stops, favStops, groups))
  }

  def predictions(stopIdsStr: String, routeIdsStr: Option[String]) = Action {
    val stopIds = stopIdsStr.split(",").map(_.toInt).toList
    val routeIds = routeIdsStr.map(_.split(",").toList).getOrElse(List())

    var predictions: Iterable[Prediction] = List()
    var msg: Option[String] = None

    try {
      predictions = controller.getPredictions(stopIds.map(i => i:java.lang.Integer), routeIds)
      msg = None
    } catch {
      case t: Throwable =>
        msg = Some(t.getMessage)
    }

    val stopNames = predictions.map(_.getStopName).toSet

    Ok(views.html.predictions(stopNames, routeIds, predictions, msg))
  }

  def saveFavoriteRoute() = Action { implicit request =>
    val routeId = saveFavoriteRouteForm.bindFromRequest.get
    controller.saveFavoriteRoute(routeId)
    Redirect(routes.Application.index())
  }

  def saveFavoriteStop() = Action { implicit  request =>
    val (stopId, routeId, direction) = saveFavoriteStopForm.bindFromRequest().get
    controller.saveFavoriteStop(stopId)
    Redirect(routes.Application.stops(routeId, direction))
  }

  def createGroup() = Action { implicit request =>
    val groupName = createGroupForm.bindFromRequest.get
    controller.createGroup(groupName)
    Redirect(routes.Application.index())
  }

  def addStopToGroup() = Action { implicit request =>
    val (stopId, groupId, routeId, direction) = addStopToGroupForm.bindFromRequest.get
    controller.addStopToGroup(groupId, stopId)
    Redirect(routes.Application.stops(routeId, direction))
  }

  def getGroup(groupId: Int) = Action {
    val stops = controller.getStopIdsForGroup(groupId)
    Ok(views.html.group(groupId, stops))
  }
}
